### ISSUE TITLE
[Feature Request] Cross Fade

### ISSUE THREAD

**Issue by Ge082** [UnicornsOnLSD/finamp#237]
What I like about PlexAmp is that once a song end by itself, there is a crossfade (Which they call "Sweet Fade) so the transition between the current song and the next song is smooth, it's very effective on Live albums because it eliminates the micro moment where there is silence that is there when the next song loads
Would be cool to have the same feature

**Comment by B16WLL** [comment]
Or (if it's possible) pre-cache the opening few seconds of the next track in the queue, and implement a capless playback option for live/mixed albums.

**Comment by rom4nik** [comment]
Isn't the playback gapless already? I've listened to some live albums and always got smooth transition between tracks without any pauses (unless the artists put them in, that is).

**Comment by waynepaulward** [comment]
It would be great if there was a smooth fade between tracks I suppose it would have to cache tracks when you start to play the album so it can turn it can start by playing the track - is this a feature that will be added …

**Comment by waynepaulward** [comment]
I should have explained this better I mean like a cross fade option so the songs fade together smoothly

**Comment by jmshrv** [comment]
This isn't implemented in just_audio due to platform support (https://github.com/ryanheise/just_audio/issues/815), it may be possible once I get around to the new player backend (#177)

**Comment by BloodyIron** [comment]
I too am interested in this for streaming, but gapless does seem to work in my brief testing with downloaded content.

**Comment by waynepaulward** [comment]
They use this feature on a few open source apps ie play sub and avsub and it works well they are open source and may help

**Comment by ryanwwest** [comment]
@BloodyIron Do you know if gapless audio worked with downloaded content _when not in offline mode_? I'm guessing Finamp is smart enough to always use the downloaded/cached version and so the seamless audio would apply, but unsure.

**Comment by jmshrv** [comment]
Yeah Finamp always defaults to downloaded content when it's available

**Comment by BobcatNoah** [comment]
Would it be possible to implement cross fade by creating two instances of the music player (It seems like it may be possible)? I'd be happy to look into it if it's not too impractical.

**Comment by Chaphasilor** [comment]
@BobcatNoah yes, that was the approach suggested by Ryan Heise over in the just_audio issue. However, I'm not sure how well thing would work, it could severely complicate things regarding pre-caching and queue handling. We would have to split the `ConcatenatingAudioSource` into two sources, one for even and one for odd indices. But this would result in parallel pre-caching, which is less efficient data and memory usage-wise.  
I'm not sure if it's worth the tradeoff...

**Comment by Chaphasilor** [comment]
Not the entire queue, just a certain duration in seconds. You can configure it in the "Audio Service" settings.

I mean, you can try to implement it, maybe it works better than expected. It's just that just_audio and audio_service tend to be rather inflexible, and in my opinion reliable playback is more important than nice transitions. If you can make it work though, why not? :)

**Comment by BobcatNoah** [comment]
I have an idea. It's crazy, but it might work and "should" be easy to implement.

For this example I’ll just use a crossfade of 4 seconds. So, 4 seconds before the first song ends, you start playing the first 4 seconds of the second song. Once those 4 seconds are over, meaning the first song is over, you skip 4 seconds into the second song. If this can be done smoothly and the next song isn’t buffering, it should sound seamless.

I started implementing this as an experiment, but I’m running into an issue. For some reason the crossfade audio doesn’t start playing until the first song ends, even if I play() it before the first song ends.

**Comment by jmshrv** [comment]
hmm, are you on iOS or Android?

**Comment by BobcatNoah** [comment]
iOS simulator. Also, if you want to see my code, I could push it to my fork.

**Comment by BobcatNoah** [comment]
I took a break from coding and I think it was just an issue with my code. I sort of fixed it, now it's just playing at the end and the beginning of the song.

**Comment by Chaphasilor** [comment]
Nice! How does it sound?

**Comment by BobcatNoah** [comment]
It's not quite working yet because I'm running into an issue with seeking. I want the player to seek x amount of seconds into the next song after the end of each song. I initially tried the code below in order to seek every time the current song changes, but I get this error. Unhandled Exception: Bad state: Cannot fire new event. Controller is already firing an event
```
_player.currentIndexStream.listen((event) async {
    await seek(crossFadeLength);
});
```

Do you have any ideas of how to achieve this seeking effect (or fix the error, but there is probably a better way to do the seeking)?

**Comment by Chaphasilor** [comment]
You might want to only do the seeking if the index actually changed, because the event is fired more often. And to get started, maybe a short delay fixes the error? Just to see if it works at all...

**Comment by BobcatNoah** [comment]
This is the best I could get it to sound.
https://github.com/jmshrv/finamp/assets/84159993/9d6e70eb-58e2-4627-81d7-9b8c50be051f

**Comment by Chaphasilor** [comment]
Awesome! Just open a PR when you're ready and I'll check it out!

**Comment by drazil100** [comment]
Is cross fade going to be optional? I prefer the current functionality but I support giving additional functionality to those who want crossfade.

**Comment by BobcatNoah** [comment]
Yes, it would be optional. However, I am no longer working on the feature.
